### ISSUE TITLE
Support customizing labels, for example, creating masks for all user queries

### ISSUE THREAD

**Issue by lsy641** [OpenRLHF/OpenRLHF#591]
When fine-tuning multi-turn messages, one practice is masking all the user queries in the label space, and just letting the model predict all the assistant tokens in the conversation which serves as a whole training instance. This will save the training time compared to segmenting a conversation into many instances.

However, in the dataset processing file, the preprocessing doesn't concern the label space and just equivalate it with the output_key content

**Comment by hijkzzz** [comment]
see https://github.com/OpenRLHF/OpenRLHF/pull/586

**Comment by lsy641** [comment]
I have checked the codes. Are they verified? I am worried about the correctness because it calculates the start and end index for each assistant message by tokenizing one message by one message. Then these indexes are applied to the tokenization of all messages. According to [huggingface guideline](https://huggingface.co/docs/trl/en/sft_trainer),   due to lack of context in response_template, the same string (”### Assistant:”) is tokenized differently. So the indexes obtained by tokenizing one by one might not aligned with the results from tokenizing the whole.
@hijkzzz

**Comment by UbeCc** [comment]
> I have checked the codes. Are they verified? I am worried about the correctness because it calculates the start and end index for each assistant message by tokenizing one message by one message. Then these indexes are applied to the tokenization of all messages. According to [huggingface guideline](https://huggingface.co/docs/trl/en/sft_trainer), due to lack of context in response_template, the same string (”### Assistant:”) is tokenized differently. So the indexes obtained by tokenizing one by one might not aligned with the results from tokenizing the whole. @hijkzzz

Thanks! We consider listing models available for this feature in `README.md`

**Comment by lsy641** [comment]
@UbeCc  I recently implemented the functions I originally hoped to have. When I was implementing, I noticed even the tokenizer was not LlaMa 2, with LlaMa 3, we still need to process the index problem carefully. Qwen tokenizer might be okay for your codes. Llama 3 chat_template will give an additional <|begin_of_text|> when the message is the first in the message list.  So separating a message list into multiple messages will cause more <|begin_of_text|>.

For your reference, the template of Llama 3 is as follows:
{% set loop_messages = messages %}{% for message in loop_messages %}{% set content = '<|start_header_id|>' + message['role'] + '<|end_header_id|>\n\n'+ message['content'] | trim + '<|eot_id|>' %}{% if loop.index0 == 0 %}{% set content = bos_token + content %}{% endif %}{{ content }}{% endfor %}{% if add_generation_prompt %}{{ '<|start_header_id|>assistant<|end_header_id|>\n\n' }}{% endif %}

**Comment by lsy641** [comment]
I suggest creating a label token list for each data sample, together with input_token and attention_mask. With the label tokens stored in the dataloader,  we will get these benefits:

1. 'prompt_ids_lens' and 'infos' in the dataset will no longer need to be maintained
2.  unify multi-turn and sing-turn scenarios, using a label token list to set ignore_index to separate input_token and predict_tokens
3.  'output_key' in the dataset will no longer need to be maintained as which span to predict is flexibly determined by the label token list.
4. remove all the hard code about label re-assignment in the sft_trainer.py
5. allow users to specify either the last assistant message as labels or all the assistant messages as labels (this mode will accelerate training speed in multi-turn data)
6. compatible to any chat_template

The only concern is that there will be some additional args introduced and some args (e.g., output_key) will be abandoned.
Do you think if it is interesting? I can submit a pr if needed @hijkzzz
